### PR TITLE
security/advancedtls: swap from deprecated pkix.CertificateList to x509.RevocationList

### PR DESCRIPTION
**Pull request by gtcooke94**
[pkix.CertificateList](https://pkg.go.dev/crypto/x509/pkix#CertificateList) is deprecated.
We have an internal wrapper around this for representing CRLs. This PR updates that wrapper to use the preferred `x509.RevocationList`.

This also replaces `x509.ParseCRL` (deprecated) with `x509.ParseRevocationList`. The former supported `PEM` input, while the latter requires `DER`, so I added a utility function `parseRevocationList` that does the PEM -> DER conversion if needed, taken from the [x509.ParseCRL](https://github.com/golang/go/blob/e2f413402527505144beea443078649380e0c545/src/crypto/x509/x509.go#L1669-L1690) implementation.

The one issue here is that `x509.RevocationList` was introduced in golang 1.19. We are still supporting 1.18. To solve this, I've put build restrictions on `crl.go` and `crl_test.go` to only build on 1.19+. Also, I also added the files `crl_deprecated.go` and `crl_deprecated_test.go`, which are identical copies to the `crl.go` and `crl_test.go` files before this PR. They have the build restriction of <go1.19, so they will be used in the 1.18 build. This change is luckily very isolated and these are the only 2 files that needed the build restrictions.

RELEASE NOTES: none

### CODE REVIEW

**Comment by rockspore** [comment]
When was this deprecated? Does it work with 1.18? Looks like the test is failing.

File: security/advancedtls/crl.go [hunk 1]
```diff
@@ -516,3 +516,34 @@ func extractCRLIssuer(crlBytes []byte) ([]byte, error) {
...
+
+// pemCRLPrefix, pemType, and parseRevocationList  come largely from here
+// x509.go:
+// https://github.com/golang/go/blob/e2f413402527505144beea443078649380e0c545/src/crypto/x509/x509.go#L1669-L1690
+// We must first convert PEM to DER to be able to use the new
+// x509.ParseRevocationList instead of the deprecated x509.ParseCRL
+
+// pemCRLPrefix is the magic string that indicates that we have a PEM encoded
+// CRL.
+var pemCRLPrefix = []byte("-----BEGIN X509 CRL")
```
Comment (rockspore) [inline security/advancedtls/crl.go:532]: This is a duplicate of line 495. The pemType below is also used in extractCRLIssuer() locally. In fact, the logic in the prefix check of that function and the new one here is the same so I think it's worth refactoring it out as a helper func.

File: security/advancedtls/crl.go [hunk 2]
```diff
@@ -516,3 +516,34 @@ func extractCRLIssuer(crlBytes []byte) ([]byte, error) {
 	}
 	return issuer, nil
 }
+
+func hasExpired(crl x509.RevocationList, now time.Time) bool {
```
Comment (rockspore) [inline security/advancedtls/crl.go:520]: It's probably better to pass the pointer here, assuming x509.RevocationList is a sizable struct.

Comment (gtcooke94) [reply security/advancedtls/crl.go:520]: Good call

File: security/advancedtls/crl.go (hunk 1 above)
Comment (gtcooke94) [reply security/advancedtls/crl.go]: Removed this and functionalized the conversion

**Comment by arvindbr8** [comment]
Hi @rockspore, Could you PTAL with the changes?

**Comment by rockspore** [comment]
The PR LGTM. The question I cannot answer is whether we are fine with dropping support of go-1.18 here. Or in other words, when is grpc-go going to drop go-1.18? I know advancedtls is experimental with its own go.mod but I'd assume we generally want the support matrix to be the same across the repo.

**Review by rockspore — APPROVED** [review]

File: security/advancedtls/crl.go [hunk 3]
```diff
@@ -516,3 +523,27 @@ func extractCRLIssuer(crlBytes []byte) ([]byte, error) {
...
+}
+
+// pemCRLPrefix, pemType, and parseRevocationList  come largely from here
+// x509.go:
+// https://github.com/golang/go/blob/e2f413402527505144beea443078649380e0c545/src/crypto/x509/x509.go#L1669-L1690
+// We must first convert PEM to DER to be able to use the new
+// x509.ParseRevocationList instead of the deprecated x509.ParseCRL
+
+// pemType is the type of a PEM encoded CRL.
+var pemType = "X509 CRL"
```
Comment (rockspore) [inline security/advancedtls/crl.go:538]: nit: This could be defined as a const. Also it could be used in line 502.

Comment (gtcooke94) [reply security/advancedtls/crl.go]: Changed

**Comment by gtcooke94** [comment]
@arvindbr8 - @rockspore took a look and thinks it's good from a security perspective, can we get a general Golang review from you?

**Comment by easwars** [comment]
> The one issue here is that x509.RevocationList was introduced in golang 1.19. We are still supporting 1.18. To solve this, I've put build restrictions on crl.go and crl_test.go to only build on 1.19+. Also, I also added the files crl_deprecated.go and crl_deprecated_test.go, which are identical copies to the crl.go and crl_test.go files before this PR. They have the build restriction of <go1.19, so they will be used in the 1.18 build. This change is luckily very isolated and these are the only 2 files that needed the build restrictions.

Would keeping the old files with their old names and adding new files with something like crl_go1.19.go reduce the diff size? As it stands, the changes are huge and are hard to review for someone who doesn't deal with certs and revocation lists on a regular basis.

**Review by easwars — APPROVED** [review]

**Pull request merged**